### PR TITLE
Add non-eager registration to dispatch autogen

### PR DESCRIPTION
**Pull request by wconstab**
Summary:
Previously, the torchscript backend would be (partially) initialized at startup.
- the dispatcher registrations would be registered,
- but other backend components would not be initialized until explicitly calling
  the backend init function

With this change, the torchscript backend is not initialized until its explicit
initialization function is called.

This enables external backends to register their own backend instead of the torchscript
backend to the same (Lazy) key.

Lands a change contributed by @antoniojkim via lazy_tensor_staging branch (#73973)

Differential Revision: D35051464

### CODE REVIEW

File: torch/csrc/lazy/ts_backend/ts_backend_impl.h [hunk 1]
```diff
@@ -0,0 +1,50 @@
+#include <torch/csrc/lazy/backend/backend_interface.h>
+
+namespace torch {
+namespace lazy {
+
+class TORCH_API TSData : public torch::lazy::BackendData {
```
Comment (bdhirsh) [inline torch/csrc/lazy/ts_backend/ts_backend_impl.h:6]: are all the changes here part of this PR? (or is there any chance you can split the eager changes into their own PR for review?)

File: aten/src/ATen/templates/RegisterDispatchKey.cpp [hunk 2]
```diff
@@ -63,7 +72,22 @@ namespace {
...
 
 TORCH_LIBRARY_IMPL(aten, ${DispatchKey}, m) {
-  ${dispatch_registrations}
+  ${cpp_namespace}::Register${BackendName}${DispatchKey}Modules = [&]() {
+
+    ${dispatch_registrations}
+
+  };
+
+  // For external backends, the EAGER_REGISTRATION macro will be set to 1,
```
[the remark this answers is not in the document]
Comment (wconstab) [reply aten/src/ATen/templates/RegisterDispatchKey.cpp:81]: yes, it was meant as in external LTC backends such as one developed out of tree.  I should probably update it to say more clearly that only the in-tree lazy torchscript backend gets special behavior, and others don't

File: aten/src/ATen/templates/RegisterDispatchKey.cpp [hunk 3]
```diff
@@ -50,6 +50,15 @@
 $dispatch_headers
 $ops_headers
 
+#ifndef EAGER_REGISTRATION
+#define EAGER_REGISTRATION 1
```
Comment (bdhirsh) [inline aten/src/ATen/templates/RegisterDispatchKey.cpp:54]: I'm a little confused by this split between `DispatchKeyNativeFunctions.h` and `RegisterDispatchKey.cpp`.

Don't we always end up `#include`ing `DispatchKeyNativeFunctions.h` in this file, so the `Register${BackendName}${DispatchKey}Modules` function will already be declared? (and the macro above will already be defined).

File: torch/csrc/lazy/ts_backend/ts_backend_impl.cpp [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+torch::lazy::BackendImplInterface* GetTSBackendImpl() {
+  static TSBackendImpl* ts_backend_impl = new TSBackendImpl();
+  return ts_backend_impl;
+}
+
+void InitTorchScriptBackend() {
+  static std::unique_ptr<BackendRegistrar> s_registrar;
+  s_registrar = std::make_unique<BackendRegistrar>(GetTSBackendImpl());
+
+  #if !EAGER_REGISTRATION
```
Comment (bdhirsh) [inline torch/csrc/lazy/ts_backend/ts_backend_impl.cpp:234]: technically this should always be set to 1 for TS, right? Maybe we should just assert it.

File: aten/src/ATen/templates/RegisterDispatchKey.cpp [hunk 5]
```diff
@@ -63,7 +72,22 @@ namespace {
...
+
+  // For external backends, the EAGER_REGISTRATION macro will be set to 1,
+  // causing the module registration function to be called at the same time
+  // it used to be called, in other words, at c++ static initialization time.
+  // For LTC backends (i.e. the lazy TorchScript backend) the EAGER_REGISTRATION
+  // macro will be set to 0 makes the c++ static initializer simply install
+  // a function pointer, and then only upon initializing the LTC backend
+  // (i.e. via a call to say init_ts_backend) is it finally invoked.
+  #if EAGER_REGISTRATION
+  ${cpp_namespace}::Register${BackendName}${DispatchKey}Modules();
```
Comment (bdhirsh) [inline aten/src/ATen/templates/RegisterDispatchKey.cpp:89]: This might just be from a different PR, but I think we technically don't need `BackendName` here? (the function name will be unique without it). Plumbing a new `BackendName` argument through the codegen for the name seems like overkill to me, but maybe there are other places in the codegen where `BackendName` actually is required.

Comment (antoniojkim) [reply aten/src/ATen/templates/RegisterDispatchKey.cpp:89]: @bdhirsh This is something I added to differentiate between backends. There is no guarantee as far as I am aware as to the order in which the symbols are statically initialized. This means that if we had just a `RegisterLazyModules` function being defined by both the TS backend and a custom backend, then there is no way of knowing who assigned the value of that function pointer last. To avoid ambiguity, I plumbed through the backend name to make things clear and unambiguous

File: aten/src/ATen/templates/RegisterDispatchKey.cpp (hunk 3 above)
Comment (antoniojkim) [reply aten/src/ATen/templates/RegisterDispatchKey.cpp:54]: Maybe in practice this might be true. But its not explicitly clear just from the template definition that this is the case. Which is why I added the `#ifndef` clause. If it happens that there is a case (either now or in the future) where the `DispatchKeyNativeFunctions.h` is not included, it would avoid a compile error here and revert back to the default behaviour.

Comment (bdhirsh) [reply aten/src/ATen/templates/RegisterDispatchKey.cpp:54]: I think the "delay registration using a lambda" feels like a niche enough use case to me that we should try to implement it with less clutter to this file. Instead of using the macros, how do you feel about unconditionally generating the `Register{...}Modules` lambda? And still using `${eager_registration}` to decide in the codegen whether or not we call it directly in the file or not (to run the registrations at load time).

The `initTSBackend()` function wouldn't need to check a macro - it would just always call the lambda to run the registrations. You might run into problems if we got the codegen wrong and accidentally registered the TS backend on startup, but you could also just add a unit test for this (assert in one of the C++ tests that there aren't any kernels registered to the Lazy key on startup).

File: aten/src/ATen/templates/RegisterDispatchKey.cpp (hunk 5 above)
Comment (bdhirsh) [reply aten/src/ATen/templates/RegisterDispatchKey.cpp:89]: yeah good point, I can't think of an easy way to get around this. (cc @ezyang if you happen to have a different opinion. Context = this PR updates the codegen'd dispatcher registrations to be able to be delayed, and registered through a lambda instead of at load time).

File: aten/src/ATen/templates/RegisterDispatchKey.cpp (hunk 3 above)
Comment (bdhirsh) [reply aten/src/ATen/templates/RegisterDispatchKey.cpp:54]: yep, that sounds reasonable.

I kind of don't want to clutter up this file with the extra macros, mostly because this is a pretty core template used in a bunch of different contexts (used for most of our eager mode dispatcher registrations too) and so anyone looking at this file for non-LTC purposes would now have to reason about the macros.

@wconstab would you rather make changes to this PR directly, or address feedback later?

Comment (wconstab) [reply aten/src/ATen/templates/RegisterDispatchKey.cpp:54]: I think we can make the changes here.  I already landed this as-is on the staging branch, but we are almost at the point of abandoning the staging branch so I think we can start to do things differently here instead of striving to stay in sync there.

It also depends how we make the changes; if you have some clear ideas of what i should change I can just make the changes.  If you actually want to try to modify it yourself, I could either land it first and you change it, or maybe you could commandeer this diff.

File: torch/csrc/lazy/ts_backend/ts_backend_impl.cpp (hunk 4 above)
Comment (wconstab) [reply torch/csrc/lazy/ts_backend/ts_backend_impl.cpp:234]: Yea, I think that's true.

File: aten/src/ATen/templates/RegisterDispatchKey.cpp (hunk 3 above)
Comment (bdhirsh) [reply aten/src/ATen/templates/RegisterDispatchKey.cpp:54]: I think what @antoniojkim put sounds reasonable:

(1) kill the macro and unconditionally generate the `Register${BackendName}${DispatchKey}Modules();`

(2) In the codegen, if the `eager_registration` argument is true, then emit the code to run the function immediately (which this PR already does), and then also set it to re-assign it to a lambda that no-ops (that way you won't accidentally call it and register things multiple times - although I think the risk of this being a problem in practice is pretty small, since nobody should ever call this function except for the torchscript one. So if this is hard to do then I think it's ok if you don't add it).

File: tools/codegen/gen_lazy_tensor.py [hunk 6]
```diff
@@ -223,15 +228,20 @@ def gen_key(func: FunctionSchema) -> Tuple[str, str]:
     assert class_name is not None
 
     # Generate nativefunction declarations
+    # Note, eager registrations is set to False for the lazy TS backend as another LTC backend
+    # may want to register their own lazy kernels instead of registering the TS ones.
+    # The registration will lazily happen when init_ts_backend is called.
     gen_dispatchkey_nativefunc_headers(fm, class_name, cpp_namespace, backend_indices,
-                                       grouped_native_functions, backend_key, autograd_key)
+                                       grouped_native_functions, backend_key, autograd_key,
+                                       backend_name, eager_registration=False)
```
Comment (henrytwo) [inline tools/codegen/gen_lazy_tensor.py:236]: @wconstab I think there's a problem here. According to the [function definition](https://github.com/pytorch/pytorch/blob/641cab9a7097d1769fccd4aef8300302ffa2fb83/tools/codegen/gen_backend_stubs.py#L204-L213) of `gen_dispatchkey_nativefunc_headers`, the positional argument after `autograd_dispatch_key` is `eager_registration`, but instead you passed in `backend_name`. Since you also set the value of `eager_registration` using a keyword, the following error occurs:

```
TypeError: gen_dispatchkey_nativefunc_headers() got multiple values for argument 'eager_registration'
```

File: aten/src/ATen/templates/RegisterDispatchKey.cpp [hunk 7]
```diff
@@ -63,7 +66,13 @@ namespace {
...
 
 TORCH_LIBRARY_IMPL(aten, ${DispatchKey}, m) {
-  ${dispatch_registrations}
+  ${cpp_namespace}::Register${BackendName}${DispatchKey}Modules = [&]() {
+
+    ${dispatch_registrations}
+
+  };
+
+  ${call_register_dispatchkey_modules};
```
Comment (ezyang) [inline aten/src/ATen/templates/RegisterDispatchKey.cpp:75]: This is pretty invasive, I'm not sure I like it.

**Comment by ezyang** [comment]
@wconstab this seems like a pretty circuitous way to do something that fundamentally should be relatively simple. The `TORCH_LIBRARY` macro exists solely because we want to run the registrations at static initialization time. If you don't need to run at static initialization time, don't bother with the macro; instead, just define a public function and use `MAKE_TORCH_LIBRARY` to get the library object which you do registrations on.

So I'd expect TORCH_LIBRARY to move into the codegen code, and then you switch between defining a function or using TORCH_LIBRARY, and then all of the extra faffing about with std::function goes away.

File: tools/codegen/gen_backend_stubs.py [hunk 8]
```diff
@@ -256,7 +263,43 @@ def gen_dispatcher_registrations(
...
+}""")
+        deferred_dispatch_registrations = deferred_template.substitute(
+            backend_name=backend_name,
+            dispatch_key=dispatch_key,
+            dispatch_registrations_body=dispatch_registrations_body
+        )
+
     fm.write_with_template(f'Register{dispatch_key}.cpp', 'RegisterDispatchKey.cpp', lambda: {
+        'static_init_dispatch_registrations': static_init_dispatch_registrations,
+        'deferred_dispatch_registrations': deferred_dispatch_registrations,
```
Comment (bdhirsh) [inline tools/codegen/gen_backend_stubs.py:302]: nit: you can probably just use one argument here instead of two in the template file (`dispatch_registrations`), since right now one of these two arguments is always an empty string.

File: tools/codegen/gen_backend_stubs.py [hunk 9]
```diff
@@ -256,7 +263,43 @@ def gen_dispatcher_registrations(
...
+        static_template = CodeTemplate("""\
+TORCH_LIBRARY_IMPL(aten, $dispatch_key, m) { 
+    $dispatch_registrations_body
+};""")
+        static_init_dispatch_registrations = static_template.substitute(
+            dispatch_key=dispatch_key,
+            dispatch_registrations_body=dispatch_registrations_body
+        )
+    else:
+        deferred_template = CodeTemplate("""\
```
Comment (bdhirsh) [inline tools/codegen/gen_backend_stubs.py:289]: minor nit: most of our codegen uses f-strings now instead of `CodeTemplate`, e.g:
```
deferred_dispatch_registrations = f"""\
TORCH_API void Register{backend_name}{dispatch_key}NativeFunctions() {{
...
```

File: tools/codegen/gen_backend_stubs.py [hunk 10]
```diff
@@ -275,16 +318,10 @@ def gen_dispatcher_registrations(
...
-                selector,
-                rocm=False,
-                cpp_namespace=cpp_namespace,
-                class_method_name=f'{class_name}'),
-            grouped_native_functions
-        )),
+        # If eager_registration is True, register the native functions
+        # for the dispatch key immediately on load. Otherwise, wait to
+        # register the functions until the generated registration func is explicitly invoked
+        'export_registration_func': "" if eager_registration else "TORCH_API",
```
Comment (bdhirsh) [inline tools/codegen/gen_backend_stubs.py:324]: is this used? (this can probably just be folded into the `dispatch_registrations` param that the codegen either spits out a function or TORCH_LIBRARY block for)

File: tools/codegen/gen_backend_stubs.py (hunk 8 above)
Comment (wconstab) [reply tools/codegen/gen_backend_stubs.py:302]: but I need to put the content in 2 different places, one place is inside anonymous namespace and the other is inside at:: namespace.  Maybe I'm missing something?

File: tools/codegen/gen_backend_stubs.py (hunk 10 above)
Comment (wconstab) [reply tools/codegen/gen_backend_stubs.py:324]: yea, this was leftover, i've removed it now.

File: tools/codegen/gen_backend_stubs.py (hunk 9 above)
Comment (wconstab) [reply tools/codegen/gen_backend_stubs.py:290]: is it bad to use CodeTemplate though? I started off using f-strings, but since I am substituting a multi-line string (with indentation) CodeTemplate actually provides value here.

Comment (bdhirsh) [reply tools/codegen/gen_backend_stubs.py:290]: I think f-strings are pretty much superior to CodeTemplate (you can do multi-line f-strings just fine, like in the codegen [here](https://github.com/pytorch/pytorch/blob/d014772b9f97523156e8d316a37b5f7e646cd247/tools/codegen/dest/register_dispatch_key.py#L86)).  It's not too big of a deal though so i don't wanna get hung up on it.

File: tools/codegen/gen_backend_stubs.py (hunk 8 above)
Comment (bdhirsh) [reply tools/codegen/gen_backend_stubs.py:303]: ah yep, i forgot about that. makes sense!

**Comment by wconstab** [comment]
hmm I seem to have totally broken this (for the TS backend - other keys seem ok)

I took the liberty of using 'MAKE_TORCH_LIBRARY_IMPL' instead of literally what @ezyang suggested, maybe I misunderstood.  It seems like either nothing is getting registered (in time?) for TS, or maybe just the fallback is not working and that's breaking virtually all tests since we fall back for some factory function.  It might matter what namespace the deferred registration function is defined in?

**Comment by ezyang** [comment]
Oops, MAKE_TORCH_LIBRARY_IMPL is the right call. I don't see anything obviously wrong, I guess I would stare at the diff between here and the previous working version

File: tools/codegen/gen_backend_stubs.py [hunk 11]
```diff
@@ -256,7 +263,44 @@ def gen_dispatcher_registrations(
...
+};""")
+        static_init_dispatch_registrations = static_template.substitute(
+            dispatch_key=dispatch_key,
+            dispatch_registrations_body=dispatch_registrations_body
+        )
+    else:
+        deferred_template = CodeTemplate("""\
+TORCH_API void Register${backend_name}${dispatch_key}NativeFunctions() {
+    auto m = MAKE_TORCH_LIBRARY_IMPL(aten, $dispatch_key);
+    $dispatch_registrations_body
```
Comment (antoniojkim) [inline tools/codegen/gen_backend_stubs.py:293]: @wconstab I think this library variable needs to be global
```suggestion
static auto m = MAKE_TORCH_LIBRARY_IMPL(aten, $dispatch_key);
TORCH_API void Register${backend_name}${dispatch_key}NativeFunctions() {
    $dispatch_registrations_body
```
I made that change locally and it appeared to work for me

**Review by antoniojkim — CHANGES_REQUESTED** [review]

File: torch/csrc/lazy/ts_backend/ts_eager_fallback.cpp [hunk 12]
```diff
@@ -149,7 +149,13 @@ void ltc_eager_fallback(
...
 }
 
-TORCH_LIBRARY_IMPL(_, Lazy, m) {
+void register_ts_ltc_eager_fallback() {
+  // Most backends use TORCH_LIBRARY_* macros which perform their dispatcher
+  // registrations at static library init time, but the lazy Torchscript backend
+  // does not since it is built in the main torch lib but not always used.
+  // In particular, if another external backend wants to register itself to the
+  // same key (Lazy), Torchscript backend must not be initialized.
+  auto m = MAKE_TORCH_LIBRARY_IMPL(_, Lazy);
```
[the remark this answers is not in the document]
Comment (wconstab) [reply torch/csrc/lazy/ts_backend/ts_eager_fallback.cpp:158]: ok great, before I just commit this, let me make sure it makes sense to me that we have 2 separate globals declared here. I guess that is equivalent to what the original code did, but I didn't understand that either.

I'm also wondering what the difference between (aten, Lazy) and (_, Lazy) is: I believe I was told the latter is how you set up a 'catchall' which is what we want for our fallback, but its a little funny that 'aten' isn't also 'catchall' since all the ops are presumably in that namespace.

cc @ezyang @bdhirsh

[the remark this answers is not in the document]
Comment (bdhirsh) [reply torch/csrc/lazy/ts_backend/ts_eager_fallback.cpp:158]: Yep, you need both of those blocks. The one in the codegen is to manually register the lazy kernel for every LTC-supported operator. And the one here is to register the Lazy-to-cpu boxed fallback.

The reason for the `_` instead of `aten` is because boxed fallbacks are meant to work "on all operators". So we have a bunch of operators under the `aten` namespace (basically everything in native_functions.yaml), but external projects are allowed to register their own custom operators to the dispatcher, usually under a different namespace.

The lazy-to-cpu fallback should work on all operators, not just the ones in `aten` (this is true for all boxed fallbacks) - which is what the underscore represents.

File: test/cpp/lazy/test_lazy_ops.cpp [hunk 13]
```diff
@@ -19,7 +19,12 @@ namespace lazy {
 
 namespace {
   // This registers the torchscript backend, without which lazy device won't work
-  torch::lazy::BackendRegistrar g_registrar(GetTSBackendImpl());
+static bool inline init_backend(){
+  torch::lazy::InitTorchScriptBackend();
+  return true;
+}
+static const bool backend_initialized = init_backend();
```
Comment (bdhirsh) [inline test/cpp/lazy/test_lazy_ops.cpp:26]: It looks like ASAN isn't happy (the ASAN test is complaining about initialization-order-fiasco: https://github.com/pytorch/pytorch/runs/5746632034?check_suite_focus=true)

I think the ordering problem might be between this static bool (which eventually calls `RegisterTorchScriptLazyNativeFunctions`), and the static library object that that function uses in the codegen: `static auto m = MAKE_TORCH_LIBRARY_IMPL(aten, $dispatch_key);`.

You might need to either access `m` through a function, or maybe not make this bool static?

**Review by bdhirsh — APPROVED** [review]
lgtm!
